File: src/FSharpAux/DateTime.fs
namespace FSharpAux



//Auxiliary functions and Wrapper for System.DateTime
module DateTime =
    
    open System

    ///Returns true when the time represented by the input is based on local time, otherwise false.
    let isLocal (d:DateTime) =
        match d.Kind with
        |DateTimeKind.Local -> true
        | _ -> false
    
    ///Returns true when the time represented by the input is based on Coordinated Universal Time (UTC), otherwise false.
    let isUtc (d:DateTime) =
        match d.Kind with
        |DateTimeKind.Utc -> true
        | _ -> false
    
    ///Returns true when the time represented by the input is based on neither local time or Coordinated Universal Time (UTC), otherwise false.
    let isUnspecified (d:DateTime) =
        match d.Kind with
        |DateTimeKind.Unspecified -> true
        | _ -> false

    ///Converts the string representation of a date and time to its DateTime equivalent. Throws an exception if the string is not in correct format.
    let parse (s:string) = DateTime.Parse(s)

    ///Converts the specified string representation of a date and time to its DateTime equivalent. Returns Some value if the conversion succeeded, and otherwise None.
    let tryParse (s:string) = 
        match System.DateTime.TryParse(s) with
        |true,res -> Some res
        |_ -> None
    
    ///Converts the value of the current DateTime object to Coordinated Universal Time (UTC).
    let toUtc (d:DateTime) = d.ToUniversalTime()

    ///Converts the value of the current DateTime object to local time.
    let toLocal (d:DateTime) = d.ToLocalTime()

    ///Compares two instances of DateTime and returns an integer that indicates whether the first instance is earlier than, the same as, or later than the second instance.
    ///Before comparing DateTime objects, ensure that the objects represent times in the same time zone.
    let compare (timeA:DateTime) (toTimeB:DateTime) =
        timeA.CompareTo(toTimeB)
    
    ///Compares two instances of DateTime and returns an integer that indicates whether the first instance is earlier than, the same as, or later than the second instance. 
    ///Performs a conversion of the two input times to the specified DateTimeKind before comparing
    let compareWithConversion (commonFormat:DateTimeKind) (timeA:DateTime) (toTimeB:DateTime) =
        match commonFormat with
        |DateTimeKind.Local -> compare (toLocal timeA) (toLocal toTimeB)
        |DateTimeKind.Utc -> compare (toUtc timeA) (toUtc toTimeB)
        | _ -> compare timeA toTimeB

    ///Compares two instances of DateTime and returns true if the first instance is earlier than the second instance, and false otherwise. 
    let isEarlier (timeA:DateTime) (toTimeB:DateTime) =
        (compare timeA toTimeB) < 0

    ///Compares two instances of DateTime and returns true if the first instance is earlier than the second instance, and false otherwise. 
    ///Performs a conversion of the two input times to the specified DateTimeKind before comparing
    let isEarlierWithConversion (commonFormat:DateTimeKind) (timeA:DateTime) (toTimeB:DateTime) =
        match commonFormat with
        |DateTimeKind.Local -> isEarlier (toLocal timeA) (toLocal toTimeB)
        |DateTimeKind.Utc -> isEarlier (toUtc timeA) (toUtc toTimeB)
        | _ -> isEarlier timeA toTimeB

    ///Compares two instances of DateTime and returns true if the first instance is later than the second instance, and false otherwise. 
    let isLater (timeA:DateTime) (toTimeB:DateTime) =
        (compare timeA toTimeB) > 0

    ///Compares two instances of DateTime and returns true if the first instance is later than the second instance, and false otherwise. 
    ///Performs a conversion of the two input times to the specified DateTimeKind before comparing
    let isLaterWithConversion (commonFormat:DateTimeKind) (timeA:DateTime) (toTimeB:DateTime) =
        match commonFormat with
        |DateTimeKind.Local -> isLater (toLocal timeA) (toLocal toTimeB)
        |DateTimeKind.Utc -> isLater (toUtc timeA) (toUtc toTimeB)
        | _ -> isLater timeA toTimeB

    ///Compares two instances of DateTime and returns true if the first instance is the same as the second instance, and false otherwise. 
    let isSame (timeA:DateTime) (toTimeB:DateTime) =
        (compare timeA toTimeB) = 0

    ///Compares two instances of DateTime and returns true if the first instance is is the same as the second instance, and false otherwise. 
    ///Performs a conversion of the two input times to the specified DateTimeKind before comparing
    let isSameWithConversion (commonFormat:DateTimeKind) (timeA:DateTime) (toTimeB:DateTime) =
        match commonFormat with
        |DateTimeKind.Local -> isSame (toLocal timeA) (toLocal toTimeB)
        |DateTimeKind.Utc -> isSame (toUtc timeA) (toUtc toTimeB)
        | _ -> isSame timeA toTimeB

File: tests/FSharpAux.Mocha.Tests/fable/src/FSharpAux.Core/DateTime.js
import { compareTo, toLocalTime, toUniversalTime, tryParse as tryParse_1, minValue, parse as parse_1, kind } from "../../fable_modules/fable-library.4.0.6/Date.js";
import { FSharpRef } from "../../fable_modules/fable-library.4.0.6/Types.js";

export function isLocal(d) {
    if (kind(d) === 2) {
        return true;
    }
    else {
        return false;
    }
}

export function isUtc(d) {
    if (kind(d) === 1) {
        return true;
    }
    else {
        return false;
    }
}

export function isUnspecified(d) {
    if (kind(d) === 0) {
        return true;
    }
    else {
        return false;
    }
}

export function parse(s) {
    return parse_1(s);
}

export function tryParse(s) {
    let matchValue;
    let outArg = minValue();
    matchValue = [tryParse_1(s, new FSharpRef(() => outArg, (v) => {
        outArg = v;
    })), outArg];
    if (matchValue[0]) {
        return matchValue[1];
    }
    else {
        return void 0;
    }
}

export function toUtc(d) {
    return toUniversalTime(d);
}

export function toLocal(d) {
    return toLocalTime(d);
}

export function compare(timeA, toTimeB) {
    return compareTo(timeA, toTimeB);
}

export function compareWithConversion(commonFormat, timeA, toTimeB) {
    switch (commonFormat) {
        case 1:
            return compare(toUtc(timeA), toUtc(toTimeB)) | 0;
        case 2:
            return compare(toLocal(timeA), toLocal(toTimeB)) | 0;
        default:
            return compare(timeA, toTimeB) | 0;
    }
}

export function isEarlier(timeA, toTimeB) {
    return compare(timeA, toTimeB) < 0;
}

export function isEarlierWithConversion(commonFormat, timeA, toTimeB) {
    switch (commonFormat) {
        case 1:
            return isEarlier(toUtc(timeA), toUtc(toTimeB));
        case 2:
            return isEarlier(toLocal(timeA), toLocal(toTimeB));
        default:
            return isEarlier(timeA, toTimeB);
    }
}

export function isLater(timeA, toTimeB) {
    return compare(timeA, toTimeB) > 0;
}

export function isLaterWithConversion(commonFormat, timeA, toTimeB) {
    switch (commonFormat) {
        case 1:
            return isLater(toUtc(timeA), toUtc(toTimeB));
        case 2:
            return isLater(toLocal(timeA), toLocal(toTimeB));
        default:
            return isLater(timeA, toTimeB);
    }
}

export function isSame(timeA, toTimeB) {
    return compare(timeA, toTimeB) === 0;
}

export function isSameWithConversion(commonFormat, timeA, toTimeB) {
    switch (commonFormat) {
        case 1:
            return isSame(toUtc(timeA), toUtc(toTimeB));
        case 2:
            return isSame(toLocal(timeA), toLocal(toTimeB));
        default:
            return isSame(timeA, toTimeB);
    }
}

